packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -3687,30 +3687,36 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
       oldEntriesBottom += 1;
     }
 
-    int pagelessRoutesToSkip = 0;
+    final List<_RouteEntry> unattachedPagelessRoutes=<_RouteEntry>[];
     // Scans the top of the list until we found a page-based route that cannot be
     // updated.
-    while ((oldEntriesBottom <= oldEntriesTop) && (newPagesBottom <= newPagesTop)) {
+    while ((oldEntriesBottom <= oldEntriesTop) && (newPagesBottom <= newPagesTop)) {    
       final _RouteEntry oldEntry = _history[oldEntriesTop];
       assert(oldEntry != null && oldEntry.currentState != _RouteLifecycle.disposed);
       if (!oldEntry.pageBased) {
-        // This route might need to be skipped if we can not find a page above.
-        pagelessRoutesToSkip += 1;
+        unattachedPagelessRoutes.add(oldEntry);
         oldEntriesTop -= 1;
         continue;
       }
       final Page<dynamic> newPage = widget.pages[newPagesTop];
       if (!oldEntry.canUpdateFrom(newPage)) {
         break;
       }
-      // We found the page for all the consecutive pageless routes below. Those
-      // pageless routes do not need to be skipped.
-      pagelessRoutesToSkip = 0;
+      // We found the page for all the consecutive pageless routes below. Attach these
+      // pageless routes to the page.
+      if(unattachedPagelessRoutes.isNotEmpty) {
```

Comment:
I think we need to add unattachedPagelessRoutes to pageRouteToPagelessRoutes even if the line 3702 evalulate to true. I think we should copy the logic in line 3667

Reply:
Add unattachedPagelessRoutes to pageRouteToPagelessRoutes even if the line 3702 evalulate to true 

>> Done

Should copy the logic in line 3667 
>> The logic is almost the same, just slightly different because it's scanning from top to bottom.

Reply:
We can do the same thing to do a putIfAbsent in the `if (!oldEntry.pageBased) {` I think? that way we will only create the list if there is pageless route.

Reply:
Because we scan from the top to the bottom, we get the pageless routes before the page route.
We cant not do` putIfAbsent in the if (!oldEntry.pageBased)  ` because we don't know the page it will attach to yet.

And we also only created the list if there is pageless route by checking `if(unattachedPagelessRoutes.isNotEmpty)`

Reply:
Add unattachedPagelessRoutes to pageRouteToPagelessRoutes even if the line 3702 evalulate to true
>> If the line 3702 `if (!oldEntry.canUpdateFrom(newPage))` evaluate to true, the oldEntry may not be in the new pages, so we don't attach pageless routes to it, and will revert the pageless routes in line 3720 and will deal it when updating the middle of the old list.